src/common/mfu_daos.c

```diff
@@ -4075,122 +4199,138 @@ static int cont_deserialize_keys(struct hdf5_args *hdf5,
-                MFU_LOG(MFU_LOG_ERR, "failed to read rx_dset");
-                rc = 1;
-                goto out;
-            }
-            hdf5->rx_dspace = H5Dget_space(hdf5->rx_dset);
-            if (hdf5->rx_dspace < 0) {
-                MFU_LOG(MFU_LOG_ERR, "failed to read rx_dspace");
-                rc = 1;
-                goto out;
-            }
-            hdf5->rx_dtype = H5Dget_type(hdf5->rx_dset);
-            if (hdf5->rx_dtype < 0) {
-                MFU_LOG(MFU_LOG_ERR, "failed to read rx_dtype");
-                rc = 1;
-                goto out;
-            }
-            hdf5->plist = H5Dget_create_plist(hdf5->rx_dset);
-            if (hdf5->plist < 0) {
-                MFU_LOG(MFU_LOG_ERR, "failed to get plist");
-                rc = 1;
-                goto out;
-            }
-            rx_ndims = H5Sget_simple_extent_dims(hdf5->rx_dspace,
-                                                 hdf5->rx_dims,
-                                                 NULL);
-            if (rx_ndims < 0) {
-                MFU_LOG(MFU_LOG_ERR, "failed to get rx_ndims");
-                rc = 1;
-                goto out;
-            }
-            num_attrs = H5Aget_num_attrs(hdf5->rx_dset);
-            if (num_attrs < 0) {
-                MFU_LOG(MFU_LOG_ERR, "failed to get num attrs");
-                rc = 1;
+        /* if rec_kv_val.len != 0 then skip akey iteration, we can
+         * write data back into DAOS using the daos_kv.h API using just
+         * oid, dkey, and key value (stored in dkey dataset) */
+
+        /* run daos_kv_put on rec_kv_val (dkey val) and key (dkey) */
+        /* skip akey iteration for DAOS_OF_KV_FLAT objects */
+        if (hdf5->dkey_data[*dk_off + j].rec_kv_val.len > 0) {
+            daos_size_t kv_single_size = 0;
+            kv_single_size = hdf5->dkey_data[*dk_off + j].rec_kv_val.len;
+            rc = daos_kv_put(*oh, DAOS_TX_NONE, 0, dkey, kv_single_size,
+                             hdf5->dkey_data[*dk_off + j].rec_kv_val.p, NULL);
+            if (rc != 0) {
+                MFU_LOG(MFU_LOG_ERR, "failed to put kv object: "DF_RC, DP_RC(rc));
                 goto out;
             }
-            if (num_attrs > 0) {
-                rc = cont_deserialize_recx(hdf5, oh, diov,
-                                           num_attrs, ak_off, k, stats);
-                if (rc != 0) {
-                    MFU_LOG(MFU_LOG_ERR, "failed to deserialize recx");
+            stats->bytes_written += kv_single_size;
+        } else {
+            for(k = 0; k < total_akeys_this_dkey; k++) {
+                memset(&aiov, 0, sizeof(aiov));
+                memset(akey, 0, sizeof(akey));
+                memcpy(akey, hdf5->akey_data[ak_off + k].akey_val.p,
+                       hdf5->akey_data[ak_off + k].akey_val.len);
+                d_iov_set(&aiov, (void*)hdf5->akey_data[ak_off + k].akey_val.p,
+                          hdf5->akey_data[ak_off + k].akey_val.len);
```

Comment:
For readability and to minimize the chance for typos, I always find it to be help to define a local variable to simplify long names, especially when the same long string is used multiple times.  Having a local variable also helps when viewing values under a debugger.

I don't know the exact type needed here, but one could simplify with something like:
```suggestion
                type* akeyval = &hdf5->akey_data[ak_off + k].akey_val;
                memcpy(akey, akeyval->p, akeyval->len);
                d_iov_set(&aiov, (void*)akeyval->p, akeyval->len);
```
It's much easier to see that all four references map to the same struct this way, and under TotalView, we can just double-click ``akeyval`` to immediate see its ``p`` and ``len`` values rather than diving through a more complicated structure.

Reply:
@adammoody yeah, indexing into those datasets was getting very long